File: Java/LabCyclePrograms/Exp6/MergeSort.java
package LabCyclePrograms.Exp6;
import java.util.Scanner;

public class MergeSort {
    public static int swapcount;
        static void mergeSort(int[] a, int low, int high){
            int mid;
            if(low < high){
                mid = (low + high)/2;
                mergeSort(a,low,mid);
                mergeSort(a,mid+1,high);
                merge(a,low,mid,high);
            }
        }
        static void merge(int a[],int low, int mid,int high){
            int[] temp_array = new int[a.length];
            int i = low;
            int j = mid+1;
            int k = low;
            while(i <= mid && j <= high){
                if(a[i] <= a[j]){
                    temp_array[k++] = a[i++];

                }
                else{
                    temp_array[k++] = a[j++];
                }
            }
            while(i<=mid)
                temp_array[k++] = a[i++];
            while(j<=high)
                temp_array[k++] = a[j++];
            for(i=low;i<=high;i++){
                if (a[i] != temp_array[i]) swapcount++;
                a[i] = temp_array[i];
            }
        }

     public static void main(String[] args) {
         int[] a = new int[20];
         Scanner s = new Scanner(System.in);
         int n;
         System.out.println("enter size of the array value: ");
         n=s.nextInt();
         System.out.println("enter the array elements");
         for(int i=0;i<n;i++)
             a[i]=s.nextInt();
         mergeSort(a,0,n-1);
         System.out.println("elements after sorting are");
         for(int i=0;i<n;i++){
             System.out.println(a[i]);
         }
         //System.out.println("The no of swaps that occurred in the sorting process are: " + swapcount);
     }
}

File: CPP/Basics/arrays/sorting/mergesort.cpp
#include<iostream>

using namespace std;

void merge(int arr[], int low, int high, int mid, int n) 
{
    int temp[n];
    int p1 = low; 
    int p2 = mid+1;
    int p3 = low;

    while(p1<=mid && p2<=high)
    {
        if(arr[p1] <= arr[p2])
        {
            temp[p3++] = arr[p1++];   
        }

        else
        {
            temp[p3++] = arr[p2++];
        }
    }

    while(p1 <= mid){
        temp[p3++] = arr[p1++];
    }

    while(p2 <= high){
        temp[p3++] = arr[p2++];
    }

    for(int i=low; i <= high; i++)
        arr[i] = temp[i];
}

void mergesort(int arr[], int low, int high, int n) 
{
    if(low<high) {
        int mid = (low + high) / 2;
        mergesort(arr, low, mid, n);
        mergesort(arr, mid+1, high, n);
        merge(arr, low, high, mid, n);
    }
}

int main() {
    int n;
    cin >> n;
    int arr[n];
    
    for(int i=0; i<n;i++) {
        cin >> arr[i];
    }
    mergesort(arr, 0, n-1, n);
    cout << "the elements after sorting are : " << endl;

    for(int i=0; i<n; i++) {
        cout << arr[i] << " ";
    }
}